graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -333,3 +343,373 @@ def _process_generic_type_definition(generic_type, schema_id, existing_scalars,
+          not of a scalar type, or stitched together fields that are of different scalar types
+    """
+    field_type_names = []
+
+    for direction, field_reference in (
+        (OUTBOUND_EDGE_DIRECTION, cross_schema_edge.outbound_field_reference),
+        (INBOUND_EDGE_DIRECTION, cross_schema_edge.inbound_field_reference),
+    ):
+        type_name = field_reference.type_name
+        field_name = field_reference.field_name
+
+        fields = type_name_to_definition[type_name].fields  # List[FieldDefinition]
+        field_type = None
+        for field in fields:
+            if field.name.value == field_name:
+                field_type = get_ast_with_non_null_stripped(field.type)
+                break
+
+        if field_type is None:  # should never happen after _check_field_reference_is_valid
+            raise AssertionError(u'Unreachable code reached. Field "{}" unexpectedly '
+                                 u'not found.'.format(field_name))
+
+        if isinstance(field_type, ast_types.ListType):
+            raise InvalidCrossSchemaEdgeError(
+                u'The {}bound field of cross schema edge "{}" gives a list, while it '
+                u'should be a single scalar'.format(
+                    direction, cross_schema_edge
+                )
+            )
+        elif isinstance(field_type, ast_types.NamedType):
+            if field_type.name.value not in scalars:
+                raise InvalidCrossSchemaEdgeError(
+                    u'The {}bound field of cross schema edge "{}" is of type "{}", which '
+                    u'is not a scalar'.format(
+                        direction, cross_schema_edge, field_type.name.value
+                    )
+                )
+        else:  # since NonNull is stripped, field_type can only be ListType or NamedType
+            raise AssertionError(
+                u'Unreachable code reached. Field has missed '
+                u'type "{}"'.format(type(field_type).__name__)
+            )
+
+        field_type_names.append(field_type.name.value)
+
+    outbound_field_type_name, inbound_field_type_name = field_type_names
+    if not _scalars_match(outbound_field_type_name, inbound_field_type_name):
+        raise InvalidCrossSchemaEdgeError(
+            u'The outbound and inbound fields of edge "{}" are of different types, '
+            u'"{}" and "{}". They are expected to be of the same scalar type.'.format(
+                cross_schema_edge, outbound_field_type_name, inbound_field_type_name
+            )
+        )
+
+
+def _scalars_match(scalar_name1, scalar_name2):
+    """Return True if the two input scalars are considered to be the same, False otherwise.
+
+    For now, two input scalars are considered to be the same if they're the same scalar, if
+    one is ID and the other is String, or if one is ID and the other is Int. This may be
+    extended in the future.
+    """
```

Comment:
nit: If you don't mind, could you add a full docstring with `Args` and `Returns` sections? Our general rule of thumb is that if the description doesn't fit on one line, it needs those sections.